### PR TITLE
Import Filecoin accounts from Ledger 

### PR DESCRIPTION
**Pull request by spylogsster**
Resolves https://github.com/brave/brave-browser/issues/19285

- Added Filecoin bridge to import accounts from Ledger devices
- Added coin type to hardware accounts

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
- this code is used yet, manual tests should be done after https://github.com/brave/brave-browser/issues/20053

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -270,6 +270,7 @@
   "devDependencies": {
     "@babel/core": "7.16.0",
     "@dump247/storybook-state": "1.6.1",
+    "@glif/filecoin-wallet-provider": "2.0.0-alpha.9",
```
Comment (diracdeltas) [inline package.json:273]: i think this should be a dependency, not a devDependency

Comment (spylogsster) [reply package.json]: done

File: components/brave_wallet_ui/common/api/hardware_keyrings.ts [hunk 2]
```diff
@@ -7,17 +7,31 @@ import { assert } from 'chrome://resources/js/assert.m.js'
...
+import FilecoinLedgerKeyring from '../hardware/ledgerjs/filecoin_ledger_keyring'
 export type HardwareKeyring = LedgerBridgeKeyring | TrezorBridgeKeyring
 
-const VendorTypes = [
-  BraveWallet.TREZOR_HARDWARE_VENDOR,
-  BraveWallet.LEDGER_HARDWARE_VENDOR
-] as const
+export enum HardwareCoins {
+  FILECOIN = 'f',
+  ETH = 'e'
```
Comment (onyb) [inline components/brave_wallet_ui/common/api/hardware_keyrings.ts:16]: Can we use more descriptive names here? Ex: `filecoin` and `ethereum`.

File: components/brave_wallet_ui/common/api/hardware_keyrings.ts [hunk 3]
```diff
@@ -7,17 +7,31 @@ import { assert } from 'chrome://resources/js/assert.m.js'
 import { BraveWallet } from '../../constants/types'
 import LedgerBridgeKeyring from '../../common/hardware/ledgerjs/eth_ledger_bridge_keyring'
 import TrezorBridgeKeyring from '../../common/hardware/trezor/trezor_bridge_keyring'
-
+import { FilecoinKeyring, LedgerKeyring } from '../hardware/hardwareKeyring'
```
Comment (onyb) [inline components/brave_wallet_ui/common/api/hardware_keyrings.ts:10]: Can we name this file `components/brave_wallet_ui/common/hardware/interfaces.ts`, so we can import it like this?

```ts
import * as HWInterfaces from '../hardware/interfaces'
```

and access it like:

```
HWInterfaces.LedgerKeyring
```

so there's no confusion with the implementations of these interfaces.

File: components/brave_wallet_ui/common/async/lib.ts [hunk 4]
```diff
@@ -48,7 +48,7 @@ export const getERC20Allowance = (
 export const onConnectHardwareWallet = (opts: HardwareWalletConnectOpts): Promise<BraveWallet.HardwareWalletAccount[]> => {
   return new Promise(async (resolve, reject) => {
     const keyring = getHardwareKeyring(opts.hardware)
-    if (keyring instanceof LedgerBridgeKeyring || keyring instanceof TrezorBridgeKeyring) {
+    if (keyring instanceof LedgerKeyring || keyring instanceof TrezorBridgeKeyring) {
```
Comment (onyb) [inline components/brave_wallet_ui/common/async/lib.ts:51]: `LedgerKeyring` is an interface, whereas I think `TrezorBridgeKeyring` is an implementation. Can we use the same for both cases?

File: components/brave_wallet_ui/common/hardware/hardwareKeyring.ts [hunk 5]
```diff
@@ -2,24 +2,30 @@
...
   abstract signPersonalMessage (path: string, message: string): Promise<SignHardwareMessageOperationResult>
 }
 
 export abstract class LedgerKeyring extends HardwareKeyring {
+  abstract getAccounts (from: number, to: number, scheme: string): Promise<GetAccountsHardwareOperationResult>
   abstract signPersonalMessage (path: string, address: string, message: string): Promise<SignHardwareMessageOperationResult>
   abstract signTransaction (path: string, rawTxHex: string): Promise<SignHardwareTransactionOperationResult>
 }
+
+export abstract class FilecoinKeyring extends HardwareKeyring {
```
Comment (onyb) [inline components/brave_wallet_ui/common/hardware/hardwareKeyring.ts:29]: Should we call this `LedgerFilecoinKeyring`? Similarly, perhaps we should rename `LedgerKeyring` to `LedgerEthereumKeyring`?

File: components/brave_wallet_ui/common/hardware/ledgerjs/filecoin_ledger_keyring.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    return ''
+  }
+  return Buffer.from(bls.getPublicKey(reversedKey)).toString('hex')
+}
+
+export function encodeKeyToHex (key: string): string {
+  return Buffer.from(key, 'base64').toString('hex')
+}
+
+export default class FilecoinLedgerKeyring extends FilecoinKeyring {
```
Comment (onyb) [inline components/brave_wallet_ui/common/hardware/ledgerjs/filecoin_ledger_keyring.ts:41]: Should we use `extends` here or `implements`?

**Review by onyb — CHANGES_REQUESTED** [review]

File: components/brave_wallet_ui/common/hardware/ledgerjs/filecoin_ledger_keyring.ts (hunk 6 above)
Comment (spylogsster) [reply components/brave_wallet_ui/common/hardware/ledgerjs/filecoin_ledger_keyring.ts]: done

File: components/brave_wallet_ui/common/hardware/hardwareKeyring.ts (hunk 5 above)
Comment (spylogsster) [reply components/brave_wallet_ui/common/hardware/hardwareKeyring.ts]: done

File: components/brave_wallet_ui/common/async/lib.ts (hunk 4 above)
Comment (spylogsster) [reply components/brave_wallet_ui/common/async/lib.ts]: done

File: components/brave_wallet_ui/common/api/hardware_keyrings.ts (hunk 3 above)
Comment (spylogsster) [reply components/brave_wallet_ui/common/api/hardware_keyrings.ts]: done

File: components/brave_wallet_ui/common/api/hardware_keyrings.ts (hunk 2 above)
Comment (spylogsster) [reply components/brave_wallet_ui/common/api/hardware_keyrings.ts]: done

**Review by darkdh — APPROVED** [review]

**Review by onyb — APPROVED** [review]

**Comment by bbondy** [comment]
Reverted here because it was causing brave://wallet to not load with an error
Screen Shot 2022-01-07 at 9 32 41 AM

**Pull request merged**